src/antsibull/collection_meta.py

```diff
@@ -13,50 +13,134 @@
 
 import os
 import typing as t
-from collections.abc import Mapping
 
+import pydantic as p
 from antsibull_fileutils.yaml import load_yaml_file
+from packaging.version import Version as PypiVer
+from pydantic.functional_validators import BeforeValidator
+from typing_extensions import Annotated, Self
 
+if t.TYPE_CHECKING:
+    from _typeshed import StrPath
 
-class CollectionMetadata:
+
+def _convert_pypi_version(v: t.Any) -> t.Any:
+    if isinstance(v, str):
+        if not v:
+            raise ValueError(f"must be a non-trivial string, got {v!r}")
+        version = PypiVer(v)
+    elif isinstance(v, PypiVer):
+        version = v
+    else:
+        raise ValueError(f"must be a string or PypiVer object, got {v!r}")
+
+    if len(version.release) != 3:
+        raise ValueError(
+            f"must be a version with three release numbers (e.g. 1.2.3, 2.3.4a1), got {v!r}"
+        )
+    return version
+
+
+PydanticPypiVersion = Annotated[PypiVer, BeforeValidator(_convert_pypi_version)]
+
+
+class RemovalInformation(p.BaseModel):
+    """
+    Stores metadata on when and why a collection will get removed.
+    """
+
+    model_config = p.ConfigDict(extra="ignore", arbitrary_types_allowed=True)
+
+    major_version: t.Union[int, t.Literal["TBD"]]
+    reason: t.Literal["deprecated", "considered-unmaintained", "renamed", "other"]
```

Comment:
One more idea: should we add `guidelines-violation` as another valid reason? I suppose this _could_ fall under `considered-unmaintained`, but there are different criteria for each under the collection removal policy IIRC.

Reply:
Since "guideline violation" should usually come with some extra explanation, I think using `other` is also fine.

Reply:
Or we add `guidelines-violation`, and enforce providing `reason_text` in that case as well.

Reply:
> Or we add `guidelines-violation`, and enforce providing `reason_text` in that case as well.

I prefer that option.

Reply:
Implemented in 2b6a5168fe49bcff98c45f06d075e309d24c3747.